File: c/examples/gpu_to_ane.m
/* gpu_to_ane.m — Zero-copy GPU → ANE handoff via shared IOSurface.
 *
 * Verifies end-to-end:
 *   1. An IOSurface can be wrapped as an MTLBuffer for GPU writes AND
 *      adopted by ane-bridge for ANE reads — same kernel object, no copy.
 *   2. Bytes Metal writes through the MTLBuffer are visible to ANE when
 *      the framework runs an identity model on the adopted surface.
 *
 * Synchronization between the GPU and ANE runs in this example is done
 * via `[MTLCommandBuffer waitUntilCompleted]` on the host side. The
 * private `_ANESharedWaitEvent` / `_ANESharedSignalEvent` interop with
 * MTLSharedEvent is exposed through the bridge API but the framework
 * currently aborts inside the eval thread when an MTLSharedEvent is
 * attached to a request — that integration path is left as future
 * work; the API surface is in place once the framework-side recipe is
 * understood.
 *
 * Build (via top-level Makefile):
 *   make examples
 *
 * Run:
 *   uv run python tools/make_identity_model.py build/identity
 *   ./build/bin/gpu_to_ane build/identity/model.mil build/identity/weights.bin
 */
#import <Foundation/Foundation.h>
#import <Metal/Metal.h>
#import <IOSurface/IOSurface.h>
#import <IOSurface/IOSurfaceRef.h>
#import "ane_bridge.h"

#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static IOSurfaceRef make_iosurface(size_t nbytes) {
    size_t W, H;
    if (nbytes == 0) {
        W = 1;
        H = 1;
    } else if (nbytes <= 16384) {
        W = nbytes;
        H = 1;
    } else {
        W = 4096;
        while (nbytes % W) {
            W /= 2;
        }
        H = nbytes / W;
    }
    NSDictionary* props = @{
        (id)kIOSurfaceWidth: @(W),
        (id)kIOSurfaceHeight: @(H),
        (id)kIOSurfaceBytesPerElement: @1,
        (id)kIOSurfaceBytesPerRow: @(W),
        (id)kIOSurfaceAllocSize: @(nbytes ? nbytes : 1),
        (id)kIOSurfacePixelFormat: @0,
    };
    return IOSurfaceCreate((__bridge CFDictionaryRef)props);
}

__attribute__((noreturn)) static void die(const char* what, AneStatus st) {
    fprintf(stderr, "%s: status=%d, last_error=%s\n", what, (int)st, ane_last_error());
    exit(1);
}

int main(int argc, char** argv) {
    if (argc < 3) {
        fprintf(stderr, "Usage: %s <model.mil> <weights.bin>\n", argv[0]);
        return 2;
    }

    @autoreleasepool {
        AneModelOpenOptions opts = {
            .mil_path = argv[1],
            .weights_path = argv[2],
            .compile_qos = ANE_QOS_DEFAULT,
        };
        AneModel* model = NULL;
        AneStatus st = ane_model_open(&opts, &model);
        if (st != ANE_OK) {
            die("ane_model_open", st);
        }

        size_t nbytes_in = ane_model_input_nbytes(model, 0);
        size_t nbytes_out = ane_model_output_nbytes(model, 0);
        size_t n_floats = nbytes_in / sizeof(float);
        if (nbytes_in != nbytes_out) {
            fprintf(stderr, "identity model expected; got in=%zu out=%zu\n", nbytes_in, nbytes_out);
            return 1;
        }

        id<MTLDevice> device = MTLCreateSystemDefaultDevice();
        if (!device) {
            fprintf(stderr, "no Metal device\n");
            return 1;
        }
        id<MTLCommandQueue> queue = [device newCommandQueue];

        IOSurfaceRef src_surface = make_iosurface(nbytes_in);
        IOSurfaceLock(src_surface, (IOSurfaceLockOptions)0, NULL);
        float* src = (float*)IOSurfaceGetBaseAddress(src_surface);
        for (size_t i = 0; i < n_floats; i++) {
            src[i] = (float)((i % 100) + 1); /* fp16-exact */
        }
        IOSurfaceUnlock(src_surface, (IOSurfaceLockOptions)0, NULL);
        id<MTLBuffer> src_mtl = [device newBufferWithBytesNoCopy:src
                                                          length:nbytes_in
                                                         options:MTLResourceStorageModeShared
                                                     deallocator:nil];

        IOSurfaceRef ane_in_surface = make_iosurface(nbytes_in);
        AneBuffer* ane_in_buf = NULL;
        st = ane_buffer_adopt_iosurface((void*)ane_in_surface, nbytes_in, &ane_in_buf);
        if (st != ANE_OK) {
            die("ane_buffer_adopt_iosurface", st);
        }
        CFRelease(ane_in_surface);

        IOSurfaceRef ane_in_back = (IOSurfaceRef)ane_buffer_iosurface_ref(ane_in_buf);
        if (ane_in_back != ane_in_surface) {
            fprintf(stderr, "iosurface_ref round-trip mismatch\n");
            return 1;
        }
        void* ane_in_base = IOSurfaceGetBaseAddress(ane_in_back);
        id<MTLBuffer> ane_in_mtl = [device newBufferWithBytesNoCopy:ane_in_base
                                                             length:nbytes_in
                                                            options:MTLResourceStorageModeShared
                                                        deallocator:nil];

        AneBuffer* ane_out_buf = NULL;
        st = ane_buffer_create_for_output(model, 0, &ane_out_buf);
        if (st != ANE_OK) {
            die("ane_buffer_create_for_output", st);
        }

        AneRequest* req = NULL;
        st = ane_request_create(model, &req);
        if (st != ANE_OK) {
            die("ane_request_create", st);
        }
        if ((st = ane_request_bind_input(req, 0, ane_in_buf)) != ANE_OK) {
            die("bind_input", st);
        }
        if ((st = ane_request_bind_output(req, 0, ane_out_buf)) != ANE_OK) {
            die("bind_output", st);
        }

        /* GPU writes through MTLBuffer that aliases ANE's input IOSurface. */
        id<MTLCommandBuffer> cb = [queue commandBuffer];
        id<MTLBlitCommandEncoder> blit = [cb blitCommandEncoder];
        [blit copyFromBuffer:src_mtl
                 sourceOffset:0
                     toBuffer:ane_in_mtl
            destinationOffset:0
                         size:nbytes_in];
        [blit endEncoding];
        [cb commit];
        [cb waitUntilCompleted];

        st = ane_request_run(req, ANE_QOS_DEFAULT);
        if (st != ANE_OK) {
            die("ane_request_run", st);
        }

        void* out_ptr = NULL;
        ane_buffer_lock(ane_out_buf, ANE_LOCK_READ, &out_ptr);
        const float* out = (const float*)out_ptr;
        size_t matches = 0, mismatches = 0;
        float max_abs_err = 0.0f;
        for (size_t i = 0; i < n_floats; i++) {
            float expected = (float)((i % 100) + 1);
            float err = fabsf(out[i] - expected);
            if (err > max_abs_err) {
                max_abs_err = err;
            }
            if (err < 1e-3f) {
                matches++;
            } else {
                mismatches++;
            }
        }
        ane_buffer_unlock(ane_out_buf);

        printf("=== GPU → ANE handoff via shared IOSurface ===\n");
        printf("model           : %s\n", argv[1]);
        printf("input bytes     : %zu\n", nbytes_in);
        printf("fp32 elements   : %zu\n", n_floats);
        printf("matches         : %zu / %zu\n", matches, n_floats);
        printf("mismatches      : %zu\n", mismatches);
        printf("max abs error   : %.6f\n", max_abs_err);
        int rc = (mismatches == 0) ? 0 : 1;
        printf("%s: GPU writes to a shared IOSurface are visible to ANE.\n",
               rc == 0 ? "PASS" : "FAIL");

        /* Bonus: attaching shared events to a direct request returns
         * `Unsupported` at the API instead of crashing. */
        AneSharedEvents* ev = NULL;
        ane_shared_events_create(&ev);
        id<MTLSharedEvent> mtl_evt = [device newSharedEvent];
        ane_shared_events_add_wait(ev, 1, (__bridge void*)mtl_evt, ANE_EVT_DEFAULT);
        AneRequest* req2 = NULL;
        ane_request_create(model, &req2);
        ane_request_bind_input(req2, 0, ane_in_buf);
        ane_request_set_shared_events(req2, ev);
        AneStatus reject = ane_request_run(req2, ANE_QOS_DEFAULT);
        if (reject == ANE_ERR_UNSUPPORTED) {
            printf("PASS: shared events on a direct request reject with UNSUPPORTED\n"
                   "      (use ane_chain_* for shared-event sync).\n");
        } else {
            printf("FAIL: expected UNSUPPORTED rejection, got status=%d\n", (int)reject);
            rc = 1;
        }
        ane_request_release(req2);
        ane_shared_events_release(ev);

        ane_request_release(req);
        ane_buffer_release(ane_in_buf);
        ane_buffer_release(ane_out_buf);
        ane_model_close(model);
        CFRelease(src_surface);
        return rc;
    }
}

File: rust/ane-bridge/examples/gpu_to_ane.rs
//! Rust dual of `c/examples/gpu_to_ane.m`.
//!
//! Drives Metal + IOSurface via the actively-maintained `objc2-*`
//! crate family — no hand-rolled `objc_msgSend`. End-to-end contract:
//! Metal blit-writes through an `MTLBuffer` aliased to the same
//! `IOSurface` that ANE adopts, ANE runs an identity model, and the
//! readback matches what Metal wrote. Also asserts that attaching
//! `SharedEvents` to a direct request returns `Unsupported` rather
//! than crashing the framework worker.
//!
//! Run:
//!   uv run python tools/make_identity_model.py build/identity
//!   cargo run --example gpu_to_ane -- \
//!     ../build/identity/model.mil ../build/identity/weights.bin

#![allow(
    clippy::allow_attributes,
    clippy::allow_attributes_without_reason,
    clippy::unwrap_used,
    clippy::expect_used,
    clippy::indexing_slicing,
    clippy::as_conversions,
    clippy::panic,
    clippy::print_stdout,
    clippy::dbg_macro,
    clippy::missing_panics_doc,
    clippy::missing_errors_doc,
    clippy::missing_assert_message,
    clippy::missing_docs_in_private_items,
    clippy::std_instead_of_core,
    clippy::std_instead_of_alloc,
    clippy::separated_literal_suffix,
    clippy::unseparated_literal_suffix,
    clippy::unreadable_literal,
    clippy::shadow_unrelated,
    clippy::shadow_reuse,
    clippy::shadow_same,
    clippy::min_ident_chars,
    clippy::float_arithmetic,
    clippy::float_cmp,
    clippy::arithmetic_side_effects,
    clippy::integer_division,
    clippy::default_numeric_fallback,
    clippy::pattern_type_mismatch,
    clippy::if_then_some_else_none,
    clippy::single_call_fn,
    clippy::needless_pass_by_value,
    clippy::let_underscore_must_use,
    clippy::let_underscore_untyped,
    clippy::redundant_pub_crate,
    clippy::semicolon_outside_block,
    clippy::semicolon_inside_block,
    clippy::semicolon_if_nothing_returned,
    clippy::print_stderr,
    clippy::clone_on_ref_ptr,
    clippy::integer_division_remainder_used,
    clippy::missing_const_for_fn,
    clippy::use_debug,
    clippy::little_endian_bytes,
    clippy::big_endian_bytes,
    clippy::deref_by_slicing,
    clippy::doc_paragraphs_missing_punctuation,
    clippy::doc_markdown,
    clippy::multiple_unsafe_ops_per_block,
    clippy::undocumented_unsafe_blocks,
    clippy::unused_trait_names,
    clippy::string_slice,
    clippy::cast_possible_wrap,
    clippy::cast_ptr_alignment,
    clippy::assertions_on_result_states,
    clippy::borrow_as_ptr,
    clippy::too_many_lines,
    clippy::unused_result_ok,
    clippy::map_with_unused_argument_over_ranges,
    clippy::ignored_unit_patterns,
    clippy::unreachable,
    clippy::decimal_literal_representation,
    clippy::single_char_pattern,
    clippy::cast_precision_loss,
    reason = "Metal interop example uses raw FFI patterns + idiomatic CLI conventions"
)]

use std::env;
use std::ffi::c_void;
use std::process::ExitCode;
use std::ptr::NonNull;

use ane_bridge::{BufferAccess, Model, OpenOptions, QoS, SharedEvents, sys};

use objc2::rc::Retained;
use objc2_core_foundation::{
    CFDictionary, CFNumber, CFRetained, CFString, kCFTypeDictionaryKeyCallBacks,
    kCFTypeDictionaryValueCallBacks,
};
use objc2_io_surface::{IOSurfaceLockOptions, IOSurfaceRef};
use objc2_metal::{
    MTLBlitCommandEncoder, MTLCommandBuffer, MTLCommandEncoder, MTLCommandQueue,
    MTLCreateSystemDefaultDevice, MTLDevice, MTLResourceOptions,
};

fn make_iosurface_props(nbytes: usize) -> CFRetained<CFDictionary> {
    let (w, h): (i64, i64) = if nbytes <= 16384 {
        (nbytes.max(1) as i64, 1)
    } else {
        let mut w = 4096i64;
        while (nbytes as i64) % w != 0 {
            w /= 2;
        }
        (w, (nbytes as i64) / w)
    };

    let entries: [(CFRetained<CFString>, CFRetained<CFNumber>); 6] = [
        (
            CFString::from_static_str("IOSurfaceWidth"),
            CFNumber::new_i64(w),
        ),
        (
            CFString::from_static_str("IOSurfaceHeight"),
            CFNumber::new_i64(h),
        ),
        (
            CFString::from_static_str("IOSurfaceBytesPerElement"),
            CFNumber::new_i32(1),
        ),
        (
            CFString::from_static_str("IOSurfaceBytesPerRow"),
            CFNumber::new_i64(w),
        ),
        (
            CFString::from_static_str("IOSurfaceAllocSize"),
            CFNumber::new_i64(nbytes.max(1) as i64),
        ),
        (
            CFString::from_static_str("IOSurfacePixelFormat"),
            CFNumber::new_i32(0),
        ),
    ];
    let mut keys: [*const c_void; 6] = [core::ptr::null(); 6];
    let mut vals: [*const c_void; 6] = [core::ptr::null(); 6];
    for (i, (k, v)) in entries.iter().enumerate() {
        keys[i] = CFRetained::as_ptr(k).as_ptr().cast();
        vals[i] = CFRetained::as_ptr(v).as_ptr().cast();
    }
    // SAFETY: `entries` owns the keys/values for the duration of the
    // call; CFDictionaryCreate retains them via the type callbacks.
    unsafe {
        CFDictionary::new(
            None,
            keys.as_mut_ptr(),
            vals.as_mut_ptr(),
            keys.len() as isize,
            &kCFTypeDictionaryKeyCallBacks,
            &kCFTypeDictionaryValueCallBacks,
        )
        .expect("CFDictionaryCreate")
    }
}

fn fill_iosurface_with_pattern(s: &IOSurfaceRef, n_floats: usize) {
    // SAFETY: lock for write before touching CPU-mapped bytes; unlock
    // after. Seed is unused (pass null).
    unsafe {
        let kr = s.lock(IOSurfaceLockOptions::empty(), core::ptr::null_mut());
        assert_eq!(kr, 0, "IOSurfaceLock failed: {kr}");
    }
    let base = s.base_address().as_ptr().cast::<f32>();
    for i in 0..n_floats {
        unsafe { base.add(i).write(((i % 100) + 1) as f32) };
    }
    unsafe {
        let kr = s.unlock(IOSurfaceLockOptions::empty(), core::ptr::null_mut());
        assert_eq!(kr, 0, "IOSurfaceUnlock failed: {kr}");
    }
}

fn iosurface_base(s: &IOSurfaceRef) -> NonNull<c_void> {
    s.base_address()
}

fn main() -> ExitCode {
    let args: Vec<String> = env::args().collect();
    if args.len() < 3 {
        eprintln!("Usage: {} <model.mil> <weights.bin>", args[0]);
        return ExitCode::from(2);
    }

    let model = Model::open(&OpenOptions::new(&args[1], &args[2])).expect("open");
    let nbytes_in = model.input_nbytes(0);
    let nbytes_out = model.output_nbytes(0);
    assert_eq!(nbytes_in, nbytes_out, "identity model expected");
    let n_floats = nbytes_in / 4;

    let device = MTLCreateSystemDefaultDevice().expect("Metal device");
    let queue = device.newCommandQueue().expect("command queue");

    // --- Source IOSurface, prefilled with fp16-exact integers ---
    let src_props = make_iosurface_props(nbytes_in);
    let src_surface = unsafe { IOSurfaceRef::new(&src_props) }.expect("IOSurfaceCreate src");
    fill_iosurface_with_pattern(&src_surface, n_floats);

    let src_mtl = unsafe {
        device.newBufferWithBytesNoCopy_length_options_deallocator(
            iosurface_base(&src_surface),
            nbytes_in,
            MTLResourceOptions::StorageModeShared,
            None,
        )
    }
    .expect("new src buffer");

    // --- ANE-side IOSurface adopted into AneBuffer, also wrapped as MTLBuffer ---
    let ane_in_props = make_iosurface_props(nbytes_in);
    let ane_in_surface = unsafe { IOSurfaceRef::new(&ane_in_props) }.expect("IOSurfaceCreate ane");
    let surface_raw: *mut c_void = CFRetained::as_ptr(&ane_in_surface).as_ptr().cast();
    let ane_in_buf =
        unsafe { ane_bridge::Buffer::adopt_iosurface(surface_raw, nbytes_in) }.expect("adopt");
    let back_ref = ane_in_buf.iosurface_ref();
    assert_eq!(back_ref, surface_raw, "iosurface_ref round-trip");

    let ane_in_mtl = unsafe {
        device.newBufferWithBytesNoCopy_length_options_deallocator(
            iosurface_base(&ane_in_surface),
            nbytes_in,
            MTLResourceOptions::StorageModeShared,
            None,
        )
    }
    .expect("new ane buffer");

    let ane_out_buf = model.output_buffer(0).expect("output buffer");

    let mut req = model.request().expect("request");
    req.bind_input(0, ane_in_buf).expect("bind_input");
    req.bind_output(0, ane_out_buf).expect("bind_output");

    // --- GPU writes through the shared IOSurface via Metal blit copy ---
    let cb = queue.commandBuffer().expect("command buffer");
    let blit = cb.blitCommandEncoder().expect("blit encoder");
    unsafe {
        blit.copyFromBuffer_sourceOffset_toBuffer_destinationOffset_size(
            &src_mtl,
            0,
            &ane_in_mtl,
            0,
            nbytes_in,
        );
    }
    blit.endEncoding();
    cb.commit();
    cb.waitUntilCompleted();

    req.run(QoS::Default).expect("ane run");

    let out_buf = req.output_buffer_mut(0).expect("output buffer bound");
    let mut matches = 0usize;
    let mut mismatches = 0usize;
    let mut max_abs_err = 0.0f32;
    out_buf
        .with_locked(BufferAccess::Read, |bytes| {
            let floats: &[f32] = unsafe {
                core::slice::from_raw_parts(bytes.as_ptr().cast::<f32>(), bytes.len() / 4)
            };
            for (i, &v) in floats.iter().enumerate() {
                let expected = ((i % 100) + 1) as f32;
                let err = (v - expected).abs();
                if err > max_abs_err {
                    max_abs_err = err;
                }
                if err < 1e-3 {
                    matches += 1;
                } else {
                    mismatches += 1;
                }
            }
        })
        .expect("output lock");

    println!("=== GPU → ANE handoff via shared IOSurface (Rust) ===");
    println!("model           : {}", args[1]);
    println!("input bytes     : {nbytes_in}");
    println!("fp32 elements   : {n_floats}");
    println!("matches         : {matches} / {n_floats}");
    println!("mismatches      : {mismatches}");
    println!("max abs error   : {max_abs_err:.6}");
    let mut rc = u8::from(mismatches != 0);
    println!(
        "{}: GPU writes to a shared IOSurface are visible to ANE.",
        if rc == 0 { "PASS" } else { "FAIL" }
    );

    // --- Safe-rejection contract: shared events on a direct request ---
    {
        let event = device.newSharedEvent().expect("shared event");
        let mut ev = SharedEvents::new().expect("shared events create");
        let mtl_evt_ptr: *mut c_void = Retained::as_ptr(&event).cast::<c_void>().cast_mut();
        unsafe { ev.add_wait(1, mtl_evt_ptr, sys::AneEventType::Default) }.expect("add_wait");

        let mut req2 = model.request().expect("request 2");
        req2.set_shared_events(Some(ev)).expect("attach events");
        match req2.run(QoS::Default) {
            Err(e) if e.status == sys::AneStatus::Unsupported => {
                println!(
                    "PASS: shared events on a direct request reject with UNSUPPORTED\n      (use ane_chain_* for shared-event sync)."
                );
            }
            other => {
                println!("FAIL: expected UNSUPPORTED rejection, got {other:?}");
                rc = 1;
            }
        }
    }

    ExitCode::from(rc)
}
